cedar-lean/Cedar/Thm/Partial/Authorization.lean

```diff
@@ -67,4 +69,103 @@ theorem partial_authz_on_concrete_gives_concrete {policies : Policies} {req : Sp
+  simp only [Partial.Response.decision]
+  simp only [PartialOnConcrete.knownForbids_eq_forbids wf]
+  simp only [PartialOnConcrete.knownPermits_eq_permits wf]
+  simp only [PartialOnConcrete.forbids_eq_satisfied_forbids wf]
+  simp only [PartialOnConcrete.permits_eq_satisfied_permits wf]
+  constructor <;> intro pid h₁
+  <;> rw [Set.in_list_iff_in_set] at *
+  <;> cases h₂ : (Spec.satisfiedPolicies .forbid policies req entities).isEmpty
+  <;> simp only [not_true_eq_false, ↓reduceIte, Bool.not_eq_true, Bool.decide_eq_false,
+    Bool.true_and, Bool.false_and, Bool.not_eq_true']
+  <;> simp only [h₂] at h₁
+  case left.false | right.false => simpa using h₁
+  case left.true | right.true =>
+    cases h₃ : (Spec.satisfiedPolicies .permit policies req entities).isEmpty
+    <;> simpa [h₃] using h₁
+
+/--
+  On concrete inputs, partial authorization's `underapproximateDeterminingPolicies`
+  are identical to concrete authorization's `determiningPolicies`
+-/
+theorem underapproximate_determining_eqv_determining_on_concrete {policies : Policies} {req : Spec.Request} {entities : Spec.Entities} {presp : Partial.Response} {resp : Spec.Response}
+  (wf : req.WellFormed) :
+  Spec.isAuthorized req entities policies = resp →
+  Partial.isAuthorized req entities policies = presp →
+  presp.underapproximateDeterminingPolicies = resp.determiningPolicies
+:= by
+  intro h₁ h₂
+  subst h₁ h₂
+  rw [← Set.eq_means_eqv Partial.Response.underapproximateDeterminingPolicies_wf determiningPolicies_wf]
+  simp only [List.Equiv, List.subset_def]
+  simp only [Partial.Response.underapproximateDeterminingPolicies, Spec.isAuthorized]
+  simp only [Partial.Response.decision]
+  simp only [PartialOnConcrete.knownForbids_eq_forbids wf]
+  simp only [PartialOnConcrete.knownPermits_eq_permits wf]
+  simp only [PartialOnConcrete.forbids_eq_satisfied_forbids wf]
+  simp only [PartialOnConcrete.permits_eq_satisfied_permits wf]
+  constructor <;> intro pid h₁
+  <;> rw [Set.in_list_iff_in_set] at *
+  <;> cases h₂ : (Spec.satisfiedPolicies .forbid policies req entities).isEmpty
+  <;> simp only [not_true_eq_false, ↓reduceIte, Bool.not_eq_true, Bool.decide_eq_false,
+    Bool.true_and, Bool.false_and, Bool.not_eq_true']
+  <;> simp only [h₂] at h₁
+  case left.false | right.false => simpa using h₁
+  case left.true | right.true =>
+    cases h₃ : (Spec.satisfiedPolicies .permit policies req entities).isEmpty
+    <;> simpa [h₃] using h₁
+
+/--
+  On concrete inputs, partial authorization's `errors` are the same policies as
+  concrete authorization's `erroringPolicies`
+-/
+theorem partial_authz_errors_eqv_erroringPolicies_on_concrete {policies : Policies} {req : Spec.Request} {entities : Spec.Entities} {presp : Partial.Response} {resp : Spec.Response}
+  (wf : req.WellFormed) :
+  Spec.isAuthorized req entities policies = resp →
+  Partial.isAuthorized req entities policies = presp →
+  Set.make (presp.errors.map Prod.fst) = resp.erroringPolicies
```

Comment:
Nit: I feel like it would be more consistent if the partial response also used the name `erroringPolicies` and returned just the policy ids.

Reply:
For Spec, we made the conscious choice that we only wanted the spec to specify which policies error, not the kind of error.  I suppose we should make the same choice here?  Just noting that this is a consequential choice in terms of how tightly we are specifying the behavior of PE.

Reply:
I feel like we probably want to make the same choice for both. Usually @emina has the strongest opinions about (over-)specifying error behavior.

Reply:
We should not over-specify error behavior :).  

The current behavior of the Spec code is at the right level of detail:  it says which policies error but does not overspecify the details.  If we mandate the kinds of errors that must returned, we will force all implementations to conform to that error spec---unless we weaken what it means to conform to the spec, which I don't think we should do without very good justification.

Reply:
I wish this was caught on #299 which introduced it :) I'll change it here though.

Reply:
Fixed and pushed